oneflow/user/kernels/dim_gather_kernels.cpp

```diff
@@ -0,0 +1,150 @@
+/*
+Copyright 2020 The OneFlow Authors. All rights reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+#include "oneflow/core/common/data_type.h"
+#include "oneflow/core/common/shape_view.h"
+#include "oneflow/core/framework/framework.h"
+#include "oneflow/user/kernels/dim_gather_kernel_util.h"
+
+namespace oneflow {
+namespace user_op {
+
+namespace {
+
+template<typename IDX_T>
+void ConvertShape2Array(const ShapeView& shape_view, IDX_T* array, int64_t num_axis) {
+  FOR_RANGE(int64_t, i, 0, num_axis) { array[i] = shape_view.At(i); }
+}
+
+}  // namespace
+
+template<DeviceType device_type, typename IN_T, typename IDX_T>
+class DimGatherKernel final : public user_op::OpKernel {
+ public:
+  DimGatherKernel() = default;
+  ~DimGatherKernel() override = default;
+
+ private:
+  void Compute(KernelComputeContext* ctx) const override {
+    const Tensor* input_tensor = ctx->Tensor4ArgNameAndIndex("input", 0);
+    const Tensor* index_tensor = ctx->Tensor4ArgNameAndIndex("index", 0);
+    Tensor* out_tensor = ctx->Tensor4ArgNameAndIndex("output", 0);
+    const int32_t dim = ctx->Attr<int32_t>("dim");
+
+    const IN_T* input = input_tensor->dptr<IN_T>();
+    const IDX_T* index = index_tensor->dptr<IDX_T>();
+    IN_T* output = out_tensor->mut_dptr<IN_T>();
+
+    int ndim = input_tensor->shape().NumAxes();
+    fixed_vector<IDX_T, kDimGatherMaxDimCount> shape_buffer(ndim);
+    auto shape2dims = [&shape_buffer, &ndim](const ShapeView& tensor_shape) {
+      std::transform(tensor_shape.ptr(), tensor_shape.ptr() + ndim, shape_buffer.begin(),
+                     [](int64_t dim) { return static_cast<IDX_T>(dim); });
```

Comment:
```suggestion
                     [](int64_t dim) -> IDX_T { return static_cast<IDX_T>(dim); });
```

lambda 尽量写清返回值类型，之前在这踩过坑。